appveyor.yml

```
@@ -119,3 +75,28 @@ artifacts:
 
 #on_finish:
 #  - ps: $blockRdp = $true; iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/appveyor/ci/master/scripts/enable-rdp.ps1'))
+
+# on build failure
+on_failure:
+- ps: |
+    Get-ChildItem -recurse artifacts\bin\tests\TestResult.xml `
+    | ForEach-Object {
+        $resultFilePath = "$_"
+        Get-Content $resultFilePath `
+        | Select-String -Pattern '<test-run id="[^"]*"\s+testcasecount="(\d+)"\s+result="([^"]*)"\s+total="(\d+)"\s+passed="(\d+)"\s+failed="(\d+)".*' `
+        | foreach {
+            $result = $_.Matches.Groups
+
+            $total  = $result[1].Value
+            $passed = $result[4].Value
+            $failed = $result[5].Value
+            $notPassed = $total - $passed
+            $testErrors = $failed
+            $testName = $resultFilePath.Split('\\')[-3]
+
+            if ($notPassed -ne 0) {
+                $logFile = Get-ChildItem -recurse artifacts\$testName*.log
+                if ($testErrors -ne 0) { Push-AppveyorArtifact $logFile }
```

The XML file contains the most complete information in one place. The `.log` file is much more readable but scattered, especially console output. The AppVeyor tab "Tests" has issues (I don't trust it).

```suggestion
            if ($notPassed -ne 0) {
                $artifactName = "TestResult.${testErrors}err.${testName}"
                $logFile = Get-ChildItem -recurse artifacts\$testName*.log
                Push-AppveyorArtifact $logFile -FileName "${artifactName}.log"
                Push-AppveyorArtifact $resultFilePath -FileName "${artifactName}.xml"
```